app/services/concerns/xccdf/profile_links.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+module Xccdf
+  # Methods related to saving rule references
+  module ProfileLinks
+    extend ActiveSupport::Concern
+
+    included do
+      def save_profile_links
+        ::ProfileLink.transaction do
+          ::ProfileLink.import!(
+            profile_links.select(&:new_record?), ignore: true
+          )
+
+          base = ::ProfileLink.joins(profile: :benchmark)
+                              .where('profiles.parent_profile_id' => nil)
+
+          profile_links_to_remove(base).delete_all
+        end
+      end
+
+      private
+
+      def profile_links
+        @profile_links ||= @op_profiles.flat_map do |op_profile|
+          profile_id = profile_id_for(ref_id: op_profile.id)
+          rule_and_rule_group_ids_for(ref_ids: op_profile.selected_rule_ids).map do |entity|
+            ::ProfileLink.find_or_initialize_by(
+              profile_id: profile_id, entity_id: entity[0], type: entity[1]
+            )
+          end
+        end
+      end
+
+      def profile_links_to_remove(base)
+        grouped_rules = profile_links.group_by(&:profile_id)
+        grouped_rules.reduce(ProfileLink.none) do |query, (profile_id, entities)|
+          query.or(
+            base.where(profile_id: profile_id)
+                .where.not(entity_id: entities.map(&:entity_id))
+          )
+        end
+      end
+
+      def profile_id_for(ref_id:)
+        @profiles.find { |p| p.ref_id == ref_id }.id
+      end
+
+      def rule_and_rule_group_ids_for(ref_ids:)
+        ref_ids.map do |ref_id|
+          rule_or_rule_group = rule_or_rule_group_for(ref_id: ref_id)
+          case rule_or_rule_group.class.name
+          when 'Rule'
+            [rule_or_rule_group&.id, 'ProfileRule']
+          when 'RuleGroup'
+            [rule_or_rule_group&.id, 'ProfileRuleGroup']
+          end
```

Comment:
I don't think you need this kind of complexity at all, this can be deleted and it can just return the object.
```suggestion
```